File: car_diary.js
// car_diary.js — JavaScript версия

const fs = require('fs');
const readline = require('readline');

const DATA_FILE = 'repairs.json';
const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

let repairs = [];

function load() {
    if (fs.existsSync(DATA_FILE)) {
        try {
            repairs = JSON.parse(fs.readFileSync(DATA_FILE, 'utf8'));
        } catch (e) {
            repairs = [];
        }
    } else {
        repairs = [];
    }
}

function save() {
    fs.writeFileSync(DATA_FILE, JSON.stringify(repairs, null, 2));
}

function ask(question) {
    return new Promise(resolve => rl.question(question, resolve));
}

function color(text, code) {
    return `\x1b[${code}m${text}\x1b[0m`;
}

async function main() {
    load();
    while (true) {
        console.log(`\n${color('🚗 Дневник автомобиля (JavaScript)', '36')}`);
        console.log("1. Добавить запись");
        console.log("2. Показать все записи");
        console.log("3. Поиск записей");
        console.log("4. Удалить запись");
        console.log("5. Статистика");
        console.log("6. Выход");
        const choice = await ask("Выберите действие: ");
        switch (choice.trim()) {
            case "1": await addRepair(); break;
            case "2": listAll(); break;
            case "3": await searchRepairs(); break;
            case "4": await deleteRepair(); break;
            case "5": stats(); break;
            case "6": console.log("До свидания!"); rl.close(); return;
            default: console.log(color("Неверный выбор.", "31"));
        }
    }
}

async function addRepair() {
    const date = await ask("Дата (ГГГГ-ММ-ДД): ");
    const mileage = parseInt(await ask("Пробег (км): "));
    const desc = await ask("Описание: ");
    const cost = parseFloat(await ask("Стоимость (руб): "));
    const repair = {
        id: repairs.length + 1,
        date: date.trim(),
        mileage: mileage,
        description: desc.trim(),
        cost: cost
    };
    repairs.push(repair);
    save();
    console.log(color(`✅ Запись добавлена (ID: ${repair.id})`, "32"));
}

function listAll() {
    if (repairs.length === 0) {
        console.log(color("Нет записей.", "33"));
        return;
    }
    console.log(color(`${'ID'.padEnd(4)} ${'Дата'.padEnd(12)} ${'Пробег'.padEnd(10)} ${'Описание'.padEnd(30)} ${'Стоимость'.padStart(10)}`, "36"));
    console.log("-".repeat(70));
    for (const r of repairs) {
        console.log(`${String(r.id).padEnd(4)} ${r.date.padEnd(12)} ${String(r.mileage).padEnd(10)} ${r.description.padEnd(30)} ${String(r.cost.toFixed(2)).padStart(10)}`);
    }
}

async function searchRepairs() {
    const keyword = (await ask("Введите ключевое слово: ")).trim().toLowerCase();
    const results = repairs.filter(r => 
        r.description.toLowerCase().includes(keyword) ||
        r.date.includes(keyword) ||
        String(r.mileage).includes(keyword)
    );
    if (results.length === 0) {
        console.log(color("Ничего не найдено.", "33"));
    } else {
        results.forEach(r => console.log(`${r.id}: ${r.date} | ${r.mileage} км | ${r.description} | ${r.cost} руб.`));
    }
}

async function deleteRepair() {
    listAll();
    const id = parseInt(await ask("Введите ID записи для удаления: "));
    const index = repairs.findIndex(r => r.id === id);
    if (index !== -1) {
        repairs.splice(index, 1);
        save();
        console.log(color("✅ Запись удалена.", "32"));
    } else {
        console.log(color("❌ Запись не найдена.", "31"));
    }
}

function stats() {
    if (repairs.length === 0) {
        console.log("Нет данных.");
        return;
    }
    const total = repairs.reduce((sum, r) => sum + r.cost, 0);
    console.log(color(`Общая стоимость ремонтов: ${total.toFixed(2)} руб.`, "32"));
    console.log(`Средняя стоимость ремонта: ${(total / repairs.length).toFixed(2)} руб.`);
    if (repairs.length > 1) {
        const intervals = [];
        for (let i = 1; i < repairs.length; i++) {
            intervals.push(repairs[i].mileage - repairs[i-1].mileage);
        }
        const avgInterval = intervals.reduce((a,b) => a+b, 0) / intervals.length;
        console.log(`Средний интервал между ремонтами: ${avgInterval.toFixed(1)} км`);
    }
}

main().catch(console.error);

File: car_diary.cs
// car_diary.cs — C# версия

using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.Json;

namespace CarDiary
{
    class Repair
    {
        public int Id { get; set; }
        public string Date { get; set; }
        public int Mileage { get; set; }
        public string Description { get; set; }
        public double Cost { get; set; }
    }

    class Program
    {
        private static List<Repair> repairs = new List<Repair>();
        private const string DataFile = "repairs.json";

        static void Main()
        {
            Load();
            while (true)
            {
                Console.WriteLine("\n\u001B[36m🚗 Дневник автомобиля (C#)\u001B[0m");
                Console.WriteLine("1. Добавить запись");
                Console.WriteLine("2. Показать все записи");
                Console.WriteLine("3. Поиск записей");
                Console.WriteLine("4. Удалить запись");
                Console.WriteLine("5. Статистика");
                Console.WriteLine("6. Выход");
                Console.Write("Выберите действие: ");
                string choice = Console.ReadLine();
                switch (choice)
                {
                    case "1": AddRepair(); break;
                    case "2": ListAll(); break;
                    case "3": SearchRepairs(); break;
                    case "4": DeleteRepair(); break;
                    case "5": Stats(); break;
                    case "6": Console.WriteLine("До свидания!"); return;
                    default: Console.WriteLine("\u001B[31mНеверный выбор.\u001B[0m"); break;
                }
            }
        }

        static void Load()
        {
            if (File.Exists(DataFile))
            {
                try
                {
                    string json = File.ReadAllText(DataFile);
                    repairs = JsonSerializer.Deserialize<List<Repair>>(json) ?? new List<Repair>();
                }
                catch
                {
                    repairs = new List<Repair>();
                }
            }
        }

        static void Save()
        {
            string json = JsonSerializer.Serialize(repairs, new JsonSerializerOptions { WriteIndented = true });
            File.WriteAllText(DataFile, json);
        }

        static void AddRepair()
        {
            Console.Write("Дата (ГГГГ-ММ-ДД): ");
            string date = Console.ReadLine();
            Console.Write("Пробег (км): ");
            int mileage = int.Parse(Console.ReadLine());
            Console.Write("Описание: ");
            string desc = Console.ReadLine();
            Console.Write("Стоимость (руб): ");
            double cost = double.Parse(Console.ReadLine());
            var repair = new Repair
            {
                Id = repairs.Count + 1,
                Date = date,
                Mileage = mileage,
                Description = desc,
                Cost = cost
            };
            repairs.Add(repair);
            Save();
            Console.WriteLine($"\u001B[32m✅ Запись добавлена (ID: {repair.Id})\u001B[0m");
        }

        static void ListAll()
        {
            if (repairs.Count == 0)
            {
                Console.WriteLine("\u001B[33mНет записей.\u001B[0m");
                return;
            }
            Console.WriteLine($"\u001B[36m{"ID",-4} {"Дата",-12} {"Пробег",-10} {"Описание",-30} {"Стоимость",10}\u001B[0m");
            Console.WriteLine(new string('-', 70));
            foreach (var r in repairs)
                Console.WriteLine($"{r.Id,-4} {r.Date,-12} {r.Mileage,-10} {r.Description,-30} {r.Cost,10:F2}");
        }

        static void SearchRepairs()
        {
            Console.Write("Введите ключевое слово: ");
            string keyword = Console.ReadLine().ToLower();
            var results = repairs.Where(r =>
                r.Description.ToLower().Contains(keyword) ||
                r.Date.Contains(keyword) ||
                r.Mileage.ToString() == keyword
            ).ToList();
            if (results.Count == 0)
                Console.WriteLine("\u001B[33mНичего не найдено.\u001B[0m");
            else
                results.ForEach(r => Console.WriteLine($"{r.Id}: {r.Date} | {r.Mileage} км | {r.Description} | {r.Cost} руб."));
        }

        static void DeleteRepair()
        {
            ListAll();
            Console.Write("Введите ID записи для удаления: ");
            int id = int.Parse(Console.ReadLine());
            var item = repairs.FirstOrDefault(r => r.Id == id);
            if (item != null)
            {
                repairs.Remove(item);
                Save();
                Console.WriteLine("\u001B[32m✅ Запись удалена.\u001B[0m");
            }
            else
                Console.WriteLine("\u001B[31m❌ Запись не найдена.\u001B[0m");
        }

        static void Stats()
        {
            if (repairs.Count == 0)
            {
                Console.WriteLine("Нет данных.");
                return;
            }
            double total = repairs.Sum(r => r.Cost);
            Console.WriteLine($"\u001B[32mОбщая стоимость ремонтов: {total:F2} руб.\u001B[0m");
            Console.WriteLine($"Средняя стоимость ремонта: {total / repairs.Count:F2} руб.");
            if (repairs.Count > 1)
            {
                var intervals = new List<int>();
                for (int i = 1; i < repairs.Count; i++)
                    intervals.Add(repairs[i].Mileage - repairs[i-1].Mileage);
                double avgInterval = intervals.Average();
                Console.WriteLine($"Средний интервал между ремонтами: {avgInterval:F1} км");
            }
        }
    }
}
